pkg/zkproofs/ciphertext_validity.go

```diff
@@ -0,0 +1,170 @@
+package zkproofs
+
+import (
+	crand "crypto/rand"
+	"encoding/json"
+	"errors"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption/elgamal"
+	"math/big"
+)
+
+// CiphertextValidityProof represents a zero-knowledge proof that a ciphertext is valid.
+type CiphertextValidityProof struct {
+	Commitment1 curves.Point
+	Commitment2 curves.Point
+	Response1   curves.Scalar
+	Response2   curves.Scalar
+}
+
+// NewCiphertextValidityProof generates a zero-knowledge proof that the given ciphertext is properly encrypted using the given Public Key.
+// Parameters:
+// - pedersenOpening: The randomness factor used in the encrypting the Ciphertext.
+// - pubKey: The public key used in the encryption.
+// - ciphertext: The ciphertext to prove the validity of.
+// - message: The message that was encrypted in the ciphertext.
+func NewCiphertextValidityProof(pedersenOpening *curves.Scalar, pubKey curves.Point, ciphertext *elgamal.Ciphertext, message uint64) (*CiphertextValidityProof, error) {
+	// Validate input
+	if pedersenOpening == nil {
+		return nil, errors.New("invalid randomness factor")
+	}
+
+	if pubKey == nil {
+		return nil, errors.New("invalid public key")
+	}
+
+	if ciphertext == nil || ciphertext.C == nil || ciphertext.D == nil {
+		return nil, errors.New("invalid ciphertext")
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+
+	H := eg.GetH()
+	G := eg.GetG()
+
+	ed25519 := curves.ED25519()
+	// Convert message to a scalar
+	messageValue := new(big.Int).SetUint64(message)
+	x, _ := ed25519.Scalar.SetBigInt(messageValue)
+
+	// Step 1: Generate random blinding factors for the proof
+	rBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value r
+	xBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value x
+
+	// Step 2: Create commitments
+	rBlindH := H.Mul(rBlind)            // rBlind * H
+	xBlindG := G.Mul(xBlind)            // xBlind * G
+	Commitment1 := rBlindH.Add(xBlindG) // Commitment1 = rBlind * H + xBlind * G
+
+	Commitment2 := pubKey.Mul(rBlind) // Commitment2 = rBlind * P
+
+	// Step 3: Generate a challenge using the Fiat-Shamir heuristic.
+	// The challenge is basically just a hash of all the provided values. This locks in the values and makes sure that
+	// the proof cannot be for some other set of values.
+	transcript := NewProofTranscript()
+	transcript.AppendMessage("C1", Commitment1.ToAffineCompressed())
+	transcript.AppendMessage("C2", Commitment2.ToAffineCompressed())
```

Comment:
Think we need to add ciphertext.C and ciphertext.D to the challenge transcript here as well to bind the proof to the specific ciphertext.

Reply:
Done